File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/LegacyGameKeyCallback.java
package in.dragonbra.javasteam.steam.handlers.steamapps.callback;

import in.dragonbra.javasteam.enums.EResult;
import in.dragonbra.javasteam.generated.MsgClientGetLegacyGameKeyResponse;
import in.dragonbra.javasteam.steam.handlers.steamapps.SteamApps;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.JobID;

/**
 * This callback is received in response to calling {@link SteamApps#getLegacyGameKey}.
 */
public class LegacyGameKeyCallback extends CallbackMsg {

    private EResult result;

    private int appID;

    private String key;

    public LegacyGameKeyCallback(JobID jobID, MsgClientGetLegacyGameKeyResponse msg, byte[] payload) {
        setJobID(jobID);
        this.appID = msg.getAppId();
        this.result = msg.getResult();

        if (msg.getLength() > 0) {
            int length = msg.getLength() - 1;
            key = new String(payload, 0, length);
        }
    }

    /**
     * @return the result of requesting this game key.
     */
    public EResult getResult() {
        return result;
    }

    /**
     * @return the appid that this game key is for.
     */
    public int getAppID() {
        return appID;
    }

    /**
     * @return the game key.
     */
    public String getKey() {
        return key;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/LegacyGameKeyCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamapps.callback

import `in`.dragonbra.javasteam.base.ClientMsg
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EResult
import `in`.dragonbra.javasteam.generated.MsgClientGetLegacyGameKeyResponse
import `in`.dragonbra.javasteam.steam.handlers.steamapps.SteamApps
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg

/**
 * This callback is received in response to calling [SteamApps.getLegacyGameKey].
 */
class LegacyGameKeyCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    /**
     * Gets the result of requesting this game key.
     */
    val result: EResult

    /**
     * Gets the appid that this game key is for.
     */
    val appID: Int

    /**
     * Gets the game key.
     */
    var key: String? = null

    init {
        val keyResponse = ClientMsg(MsgClientGetLegacyGameKeyResponse::class.java, packetMsg)
        val msg = keyResponse.body

        jobID = keyResponse.targetJobID
        appID = msg.appId
        result = msg.result

        if (msg.length > 0) {
            val length: Int = msg.length - 1
            val payload = keyResponse.payload.toByteArray()
            key = String(payload, 0, length)
        }
    }
}
